text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+2. Another option would be to use the tip of one of the forks of the chain, if more than one fork has been created before the detection of the equivocation catastrophe.
+3. The last option would be to use a manually constructed block (more details below).
+
+A desirable goal here would be to revert as few transactions from all the forks as possible, in order to limit the damage. Every transaction that was included in a block which appeared to be finalized for a time could have caused real world consequences (like, for example, a merchant sending ordered goods). Reverting any such transaction might potentially cause a loss to a user.
+
+Option 1 essentially reverts all transactions that happened after the block whose FTT has been exceeded. Option 2 attempts to revert the transactions from all forks but one. However, it is possible that some of the transactions on different forks of the chain don't conflict with each other. In such a case, we could attempt to replay all of them on the new trusted branch. This is the idea behind option 3: to gather the non-conflicting transactions, create a block containing them and collectively declare it correct, outside of the consensus algorithm.
+
+After the users collectively decide upon the last correct block, the transactions to be replayed and the list of equivocators to be slashed, the network can be restarted.
+
+### Restarting the network
+
+The process of restarting the network could be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators, which is based on the bonded validators of the last correct block, with the equivocators removed.
+- There is a different initial global state: it is not empty anymore, but it is the post-state of the last correct block.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
```

Comment:
Actually, we might want to consider more than just the validators to be the "users" here. (Vlad Zamfir made that point a few months ago.)

Social consensus can't be defined precisely, of course, but e.g. in a scenario where _all_ validators go bad, others could still decide to continue running the system, with a completely new set of validators. Or if 90% of validators go bad, others who already have bids in the auction contract could be added to the new validators. Or even if 50% equivocated, the community — the users, exchanges, etc. — might not trust the other 50% anymore and want to continue running with a different set of validators.

So maybe point 3 should allow, in addition to removing equivocators, to:
* remove validators without slashing them, and
* add new validators, and
* possibly even adding new stakes to the auction contract?

Reply:
Good point. Completely redefining the set of validators should probably be possible. This actually would make the restart after an equivocation catastrophe even more similar to the initial launch of the network, where we also define an arbitrary set of validators.

Maybe it would even be best to define handling of equivocation catastrophes in terms of a generalization of the initial network startup process?

Reply:
> Maybe it would even be best to define handling of equivocation catastrophes in terms of a generalization of the initial network startup process?

I like this idea. We could simply restart network from the "last trusted" era and use the new chainspec.

Reply:
I don't think non-validator users should be explicitly involved (big dApps might be validators, anyway), but we need to set fairly precise expectations about what happens when a catastrophe hit, and let users write their own CEPs if they want to challenge it.

Reply:
Sure, it's not feasible to have a conference with tens of thousands of users. But if the meeting is among validators only and behind closed doors, that could hurt trust in the platform even more. So the expectation should be — and maybe that's something we should explicitly recommend — that the whole process and discussion is made public.

But from the _technical_ side, I think we also need to provide the means to overcome the absolute worst-case scenario: That all validators (and we ourselves) turn evil, and the community wants to replace them. I.e. some way to continue the network from a particular state, but with completely different validators and maybe some other state changes (slash everyone!) applied.

Reply:
Well, the process can be made public without allowing public to actively participate in the _main_ communication channel.

The worst-case scenario could be provided for, but mostly it's a technical issue of restarting from a given state. We should have the feature, but we don't need to provision a governance process for a total schism between validators and the userbase. Something must have gone wrong far upstream of that scenario.